hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/statemachine/background/BlockDeletingService.java

```diff
@@ -556,17 +556,26 @@ private Pair<Integer, Long> deleteTransactions(
             LOG.warn("Missing delete block(Container = " +
                 container.getContainerData().getContainerID() + ", Block = " +
                 blkLong);
+            blocksDeleted++;
```

Comment:
I do not think we should increment this when block is not found in the container.

Reply:
When SCM sends the DeletedBlocksTransaction to DN, Dn will increase the `numPendingDeletionBlocks `.
When DN deletes the block, DN then decreases the `numPendingDeletionBlocks `.
If we do not count these unrecorded as deleted, `numPendingDeletionBlocks ` of container data will stay still and obstruct the blockDeletingService(if we use the default TopN strategy). Cuz `numPendingDeletionBlocks ` is counted into the threshold but it will not reduce. If such invalid pending blocks accumulate and exceed the threshold, no more normal container would be chosen to delete blocks.


The unrecorded block should be counted as deleted here. Since Dn now will not act as the client and actively add the unknown blocks to its DB. That's weird.

The blocks that may exist but are not recorded in the RDB could be orphan blocks created by the client crush or met an error(for. example containerNotOpenException) before **putBlock**. I am about to set up a background service to clean up them.

Reply:
Also, you could take a look at https://github.com/apache/ozone/pull/3681, that's a similar problem. These services have the common risk that the computation of the threshold counter is of great importance. Any mistake could lead to the cluster not working.

Reply:
I see. Thanks for the explanation @Xushaohong ! Good find.